src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt

```diff
@@ -1,3 +1,166 @@
+                            .any { selectedVector -> embeddings.cosine(score.vector, selectedVector) > simThreshold }
+                    }
+                    .firstOrNull()
+                    ?.let { maxScore ->
+                        selectedVectors.add(maxScore.vector)
+                        selectedIndices.add(maxScore.i)
+
+                        maxScore
+                    }
+            }
+    }
+
+    /**
+     * This scoring compares using the following:
+     * 1. Don't have a vector to similar to any vector in the current summary
+     * 2. Create centroid of currSummary + currSen and compare to centroid of whole doc
+     */
+    private fun scoreGhalandari(
+        numSentences: Int,
+        centroid: Array<Float>,
+        scoreHolders: List<ScoreHolder>
+    ): List<ScoreHolder> {
+        val selectedIndices = mutableSetOf(scoreHolders.first().i)
+        val selectedVectors = mutableSetOf(scoreHolders.first().vector)
+        var centroidSummary = scoreHolders.first().vector
+
+        return listOf(scoreHolders.first()) + (2..min(numSentences, scoreHolders.size))
+            .mapNotNull { _ ->
+                scoreHolders
+                    .filterNot { selectedIndices.contains(it.i) }
+                    .filterNot { score ->
+                        selectedVectors.any { selectedVector ->
+                            embeddings.cosine(score.vector, selectedVector) > simThreshold
+                        }
+                    }
+                    .maxBy { score -> embeddings.cosine(centroid, (score.vector + centroidSummary).normalize()) }
+                    ?.let { maxScore ->
+                        centroidSummary =
+                            (centroidSummary + maxScore.vector) // No need to normalize as it's done in the maxBy.
+                        selectedVectors.add(maxScore.vector)
+                        selectedIndices.add(maxScore.i)
+
+                        maxScore
+                    }
+            }
+    }
+
+    override fun summarize(text: String, lines: Int): String {
+        val sentences = text.sentences()
+        val superCleanSentences = sentences.map { it.simplify() }
+        val (tfIdf, wordMap) = getTfIdf(superCleanSentences)
+        val wordsOfInterest = tfIdf
+            .mapIndexedNotNull { index, d -> if (d > threshold) wordMap[index] else null }
+            .toSet()
+        val centroidVector = getWordVector(superCleanSentences.flatMap { it.words() }, wordsOfInterest)
+        val scores = getSentenceBaseScoring(superCleanSentences, sentences, centroidVector, wordsOfInterest)
+        val finalSentences = scoreGhalandari(lines, centroidVector, scores)
+
+        return finalSentences.sortedBy { it.i }.joinToString("\n") { it.rawSentence }
+    }
+
+    override fun summarize(text: String, ratio: Double): String {
+        val lines = text.sentences().size
+
+        return summarize(text, (lines * ratio).roundToInt().coerceAtLeast(1))
```

Comment:
This `(lines * ratio)` is used in more than one place, perhaps extract?

Reply:
Nah